021-bucket-folders.md

```diff
@@ -0,0 +1,223 @@
+# Folder Support for storage Buckets
+
+- Implementation Owner: @loks0n
+- Start Date: 06-10-2025
+- Target Date: TBD
+- Appwrite Issue: TBD
+
+## Summary
+
+[summary]: #summary
+
+Add folder support to Appwrite Storage for organizing files. Folders are lightweight organizational containers - they don't have independent permissions, don't support nesting initially, and exist purely to help users sort and filter files within buckets.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Large buckets become unmanageable. Users with hundreds or thousands of files have no way to organize them. Every file appears in a flat list. Finding specific files requires searching by name or using complex filtering.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Design decisions
+
+1. **Folders are organizational** - No independent permissions, compression, or encryption settings
+2. **Flat structure initially** - No nested folders (folders can't contain folders)
+3. **Backward compatible** - Existing files stay at root level
+4. **Minimal overhead** - Folders are lightweight documents, not filesystem constructs
+
+### API Endpoints
+
+**POST /v1/storage/buckets/:bucketId/folders**
+Create a folder.
+```
+bucketId: string (required)
+folderId: string (required) - Custom ID or ID.unique()
+name: string (required) - Folder name (max 255 chars)
+permissions: array (optional) - Inherits bucket permissions if null
+```
+
+**GET /v1/storage/buckets/:bucketId/folders**
+List folders in a bucket.
+```
+bucketId: string (required)
+queries: array (optional) - Standard query support
+search: string (optional)
+```
+
+**GET /v1/storage/buckets/:bucketId/folders/:folderId**
+Get folder by ID.
+
+**PUT /v1/storage/buckets/:bucketId/folders/:folderId**
+Update folder name/permissions.
+
+**DELETE /v1/storage/buckets/:bucketId/folders/:folderId**
+Delete folder. Fails if folder contains files unless `force=true`.
+
+**Modified: POST /v1/storage/buckets/:bucketId/files**
+Add `folderId` parameter (optional, defaults to null = root level).
```

Comment:
should this really be folder ID? maybe path for better DX? we can generate the hierarchy on creation or simpler throw 404 if path doesn't exists and send to create a folder..